Fix regression - Delete .terraform.lock.hcl  if it was created by terraform init

gezb opened the pull request:
This PR fixes a regression I introduced with the changes in https://github.com/runatlantis/atlantis/pull/1651 reported in https://github.com/runatlantis/atlantis/issues/1698

My initial change did not take into account that the terraform init ran as part of the workflow will create the `.terraform.lock.hcl` lock file and the knock on affects of this file existing when the project is re-planned

If the lock file exists after terraform init has run and it did not exist initially we delete it

nishkrishnan commented on server/events/project_command_runner.go:
```diff
@@ -313,11 +314,25 @@ func (p *DefaultProjectCommandRunner) doPlan(ctx models.ProjectCommandContext) (
 		if unlockErr := lockAttempt.UnlockFn(); unlockErr != nil {
 			ctx.Log.Err("error unlocking state after plan error: %v", unlockErr)
 		}
+		if !lockFileExistsProirToPlan && common.FileExists(lockfilePath) {
+			// if the lock file did not exist before running the plan delete the lock file that was generated
+			// by terraform init so that on subsequent actions modules and providers are not locked to the versions
+			// detected as a part of this plan
+			ctx.Log.Debug("Deleting `%s` that was generated by terraform init", lockFileName)
+			delErr := os.Remove(lockfilePath)
+			if delErr != nil {
+				ctx.Log.Info("Error Deleting `%s`", lockFileName)
+			}
+		}
```
Is there a reason this cleanup logic also can't live in the `InitStepRunner`?

gezb replied:
I tried that originally but that caused issues with the e2e tests

nishkrishnan replied:
What issues?

gezb replied:
if `terraform.lock.hcl` does not exist when terraform plan is run it errors with

```shell
│ Error: Could not load plugin
│ 
│ 
│ Plugin reinitialization required. Please run "terraform init".
│ 
│ Plugins are external binaries that Terraform uses to access and manipulate
│ resources. The configuration provided requires plugins which can't be located,
│ don't satisfy the version constraints, or are otherwise incompatible.
│ 
│ Terraform automatically discovers provider requirements from your
│ configuration, including providers used in child modules. To see the
│ requirements and constraints, run "terraform providers".
│ 
│ failed to instantiate provider "registry.terraform.io/hashicorp/null" to obtain schema: unknown provider "registry.terraform.io/hashicorp/null"
│ 
```

nishkrishnan commented on server/events/project_command_runner.go:
```diff
@@ -313,11 +314,25 @@ func (p *DefaultProjectCommandRunner) doPlan(ctx models.ProjectCommandContext) (
 
+	lockFileName := ".terraform.lock.hcl"
+	lockfilePath := filepath.Join(projAbsPath, lockFileName)
+	lockFileExistsProirToPlan := common.FileExists(lockfilePath)
```
```suggestion
	lockFileExistsPriorToPlan := common.FileExists(lockfilePath)
```

gezb replied:
Fixed

nishkrishnan commented on server/events/project_command_runner.go:
```diff
@@ -313,11 +314,25 @@ func (p *DefaultProjectCommandRunner) doPlan(ctx models.ProjectCommandContext) (
 		if unlockErr := lockAttempt.UnlockFn(); unlockErr != nil {
 			ctx.Log.Err("error unlocking state after plan error: %v", unlockErr)
 		}
+		if !lockFileExistsProirToPlan && common.FileExists(lockfilePath) {
```
```suggestion
		if !lockFileExistsPriorToPlan && common.FileExists(lockfilePath) {
```

gezb replied:
Fixed

nishkrishnan commented on server/events/project_command_runner.go:
```diff
@@ -313,11 +314,25 @@ func (p *DefaultProjectCommandRunner) doPlan(ctx models.ProjectCommandContext) (
 			ctx.Log.Err("error unlocking state after plan error: %v", unlockErr)
 		}
+		if !lockFileExistsProirToPlan && common.FileExists(lockfilePath) {
+			// if the lock file did not exist before running the plan delete the lock file that was generated
```
Why is this only being run when there is an error in the steps?  Does this file need to be present for an apply?  Maybe it might make sense to use something like git to detect whether this is a new file that's been generated previously and clean this up before the init runs in the step runner itself.

gezb replied:
Good spot I have moved this outside the `if err != nil {`

gezb commented:
I would like to add a test for this in the e2e tests but this would require upgrading the version of terraform used for the e2e tests to 0.14 or greater, What would be involved in doing this?

gezb commented:
this still is not correct - Looking at the CircleCI output the apply step of the e2e tests now fails because the lock file does not exist - not sure of the best way to solve this now? 

@nishkrishnan  any ideas?

nishkrishnan commented:
> this still is not correct - Looking at the CircleCI output the apply step of the e2e tests now fails because the lock file does not exist - not sure of the best way to solve this now?
> 
> @nishkrishnan any ideas?

I mentioned an idea in an above comment: 

"Maybe it might make sense to use something like git cli to detect whether this is a new file that's been generated previously and clean this up before the init runs in the step runner itself."

If the file is committed, it won't show up when running git status.

gezb commented:
OK I have rewritten this and I think its ready for re-review - One change I have had to make is require TF  > 0.14.0 in the E2E tests as I have added a test around the lock file handling

nishkrishnan commented on server/events/working_dir.go:
```diff
@@ -307,6 +309,20 @@ func (w *FileWorkspace) cloneDir(r models.Repo, p models.PullRequest, workspace
 }
 
+func (w *FileWorkspace) IsFileTracked(log logging.SimpleLogging, cloneDir string, filename string) (bool, error) {
+	cmd := exec.Command("git", "ls-files", filename)
+	cmd.Dir = cloneDir
+
+	output, err := cmd.CombinedOutput()
+
+	if err != nil {
+		log.Warn("Error checking if %s is tracked: %s", filename, string(output))
+		return false, err
+	}
+	return len(output) > 0, nil
```
I'm more in favor with just keeping all this logic in `core/runtime/common` or even in `InitStepRunner` 
1) I dont know if there is really going to be something else that uses this logic. The long term solution is to just clean up after each operation
2). `ProjectCommandContext` is becoming a dumping ground for things we need to plumb through the service and is losing it's meaning, let's not add to that unnecessarily.
3) We can leave `ProjectCommandRunner` out of this, which already has a lot going on.

gezb replied:
I have made the suggested change and moved the code to check if file is tracked in git into `core/runtime/common`

nishkrishnan reviewed (commented):
Nice! We're getting there!

nishkrishnan reviewed (approved):
💯  Going to leave it open for a few days for other maintainers to look at before merging.
